### ISSUE TITLE
Deluge and SABnzbd postprocess scripts causing sonarr and radarr to crash

### ISSUE THREAD

**Issue by Smalfinn** [mdhiggins/sickbeard_mp4_automator#853]
I cannot see any reason why this is happening but when a few downloads are being processed sonarr and or radarr crash and I have to restart their services.

Log: https://pastebin.com/bxNr6aZi

**Comment by Smalfinn** [comment]
`2018-03-06 12:46:17 - SonarrPostConversion - INFO - queued
2018-03-06 12:46:22 - SonarrPostConversion - INFO - queued
2018-03-06 12:46:27 - SonarrPostConversion - INFO - queued
2018-03-06 12:46:32 - SonarrPostConversion - INFO - queued
2018-03-06 12:46:37 - SonarrPostConversion - INFO - queued
2018-03-06 12:46:42 - SonarrPostConversion - INFO - queued
2018-03-06 12:46:47 - SonarrPostConversion - INFO - queued
2018-03-06 12:46:52 - SonarrPostConversion - INFO - queued
2018-03-06 12:46:57 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:02 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:07 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:12 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:17 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:22 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:27 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:32 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:37 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:42 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:47 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:52 - SonarrPostConversion - INFO - queued
2018-03-06 12:47:57 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:02 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:07 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:12 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:17 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:22 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:27 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:32 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:37 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:42 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:47 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:52 - SonarrPostConversion - INFO - queued
2018-03-06 12:48:57 - SonarrPostConversion - INFO - queued
2018-03-06 12:49:02 - SonarrPostConversion - INFO - queued
2018-03-06 12:49:07 - SonarrPostConversion - INFO - queued
2018-03-06 12:49:12 - SonarrPostConversion - INFO - queued`

Seems to cause Sonarr to crash

**Comment by Smalfinn** [comment]
I contacted sonarr to see if they could identify the issue: https://forums.sonarr.tv/t/i-have-to-keep-restarting-sonarr-windows-service/17572

Their answer was that the automator isn't terminating.

> "...which means the custom script never terminates so Sonarr can never process the next download. The script needs to terminate after running, not sit waiting for something to be queued."

Is this something you can look into?

**Comment by mdhiggins** [comment]
The script is waiting for Sonarr to finish its processing to go from queued to completed. It’s looping continuously. 

It triggers a rescan then waits for it to finish before proceeding to the next step. I could add a timeout so this doesn’t freeze things indefinitely but this queue process should happen asynchronously so it is actually a problem with a Sonarr because the script isn’t actually freezing it’s just persistently waiting for the RescanSeries function to complete

I’m not sure why the RescanSeries option is taking so long but I would check the Sonarr logs and maybe clarify this with the Sonarr message boards. It’s working fine on my end using Sonarr

**Comment by Smalfinn** [comment]
Would you mind running me through how you've got Sonarr and the script setup?

**Comment by Smalfinn** [comment]
markus101 from Team Sonarr in reply to

> It triggers a rescan then waits for it to finish before proceeding to the next step. I could add a timeout so this doesn’t freeze things indefinitely but this queue process should happen asynchronously so it is actually a problem with a Sonarr because the script isn’t actually freezing it’s just persistently waiting for the RescanSeries function to complete

That scan doesn’t take hours to complete, it should complete within a few minutes even if it’s trying to import hundreds of episodes.

**Comment by Rhainland** [comment]
@mdhiggins I am also experiencing this issue, currently i cant process any files in sonarr because the script "hangs".

Here are my sonarr logs;
debug = https://docs.google.com/document/d/1n3UbwPyLdhQNt8CLHAWgvtiwWmEGaM1Srf1e-FG7MJA/edit?usp=sharing

trace 1 = https://docs.google.com/document/d/1BLQHcKdlK0RFsZFHHNXLPVqswzK0X-sPSkZa1CtAOmc/edit?usp=sharing

Trace 2 = https://docs.google.com/document/d/1kTwjjqQSUj5cov-quIHbGIydsTFYnmZ2MmJj1nAs4sM/edit?usp=sharing

**Comment by Rhainland** [comment]
@mdhiggins So it seems that the python.exe processes that are spawned when a file needs to be moved, and qtfaststart is being run, are not being killed once the process is done.

I can sit here and look at task manager spawn 3(and sonarr logs indicates 3 files are now being moved) it starts qtfasttart and when the process is done, the processes are hanging and staying a live preventing the next files from being processed

**Comment by Smalfinn** [comment]
@Rhainland @mdhiggins 
Same here, starts 3 files then begins to hang - if I crash the processes sonarr kicks back into action!

**Comment by Smalfinn** [comment]
@Rhainland @mdhiggins 
Okay so I changed the autoProcess.ini to relocate_moov = False and now all works perfectly?

Must be an issue with qtfaststart!

**Comment by Rhainland** [comment]
@Smalfinn I tried this too earlier, but the issue still came back.

**Comment by Smalfinn** [comment]
@Rhainland Im 20 files deep and working perfectly

Make sure you have Completed Download Handling disabled.
http://prntscr.com/io5p98

**Comment by Rhainland** [comment]
@Smalfinn I do have completed download handling disabled, right now it seems okay but im sure its gonna come back after a while ^^

**Comment by Smalfinn** [comment]
@Rhainland HAVE FAITH!

**Comment by Rhainland** [comment]
@Smalfinn @mdhiggins Even if it does in fact work(despite having tried with it off and i experienced it didnt) i still think we need some input from mdhiggins on this so we can get the problem resolved, i would like to be able to run qtfasttart

**Comment by Smalfinn** [comment]
@Rhainland @mdhiggins Agreed

**Comment by Rhainland** [comment]
@Smalfinn Still happens if you download a lot of files really quickly, this time i downloaded about 15 files really quickly after eachother, they didnt need to be converted but just tagged and move, which spawned 4-5 python processed, that now wont shutdown on their own :(

**Comment by Smalfinn** [comment]
@Rhainland I downloaded 168 episodes and im still processing strong, no problems yet

**Comment by mdhiggins** [comment]
So this has nothin to do with QTFT or moving etc

All to these issues are stemming from a recent change which can be referenced here #849 
The goal was to attempt to have Sonarr rescan the series and then remonitor the episode after processing which wouldn't happen in certain scenarios

To do that, the postSonarr and postRadarr files were updated to talk to the API and trigger rescans, wait for completion, then remark the episode as monitored

The offending portion of code that's causing these loops is here:

```
                while command['state'] not in ['complete', 'completed']:
                    log.info(str(command['state']))
                    time.sleep(5)
                    r = requests.get(url, headers=headers)
                    command = r.json()
```

For whatever reason Sonarr isn't completing as it should, maybe its crashing or freezing or maybe there's too many requests, who knows, the logs don't indicate why the RescanSeries functionality never completes

With regards to QTFastStart, if you're tagging and moving QT fast start can add a lot of time to processing that would otherwise take seconds so its possible that those python processes that were spawning were actually doing work and not actually faulty

I went ahead and pushed some changes that add a timeout feature that will stop waiting for the queued to complete state change after 60 seconds and also increased the delay between repinging Sonarr from 5 to 10 seconds

867646bf0622f23af0ba7f8fcd83a44dec9c1106

See if that helps

**Comment by Smalfinn** [comment]
Shall I enable QTFT again or just leave it?

**Comment by mdhiggins** [comment]
I would keep it on if its a feature you want, it shouldn't have any impact on the above issues

**Comment by Smalfinn** [comment]
Is the time spent processing QTFT worth the streaming improvement? 

Probably a situational question.

**Comment by Rhainland** [comment]
@mdhiggins Thank you, i will test this right now. I added a temporary solution which constist of a batch file that kills any Python.exe if no ffmpegs are in the process list.

The python.exe's are in fact doing work, but after about a minute or so, they arent. I tested this by instantly killing the 3 python.exes that spawned, this resulted in the show not being imported, sonarr instead just deleted it from activity and redownloaded, but if i gave it 1-2 minutes after they spawned, then killed them, sonarr correctly moved the files.

So the batch file solved the problem, even though its not pretty. 

Im testing the new update and hopefully its all good!
